### PR TITLE
Matrix inverse using Block inversion

### PR DESCRIPTION
**Pull request by abh2k**
* matrices
  * Faster Matrix Inverse using Block Matrix

### CODE REVIEW

**Comment by abhinav28071999** [comment]
@sylee957 @oscarbenjamin  I was wondering that why `_verify_invertible(M, iszerofunc=iszerofunc)`. It appears to be slower than to actually calculate the inverse of the Matrix when the matrix is invertible

**Comment by sylee957** [comment]
The reason why invertibility test is slower than computing the inverse, is that it’s computing the determinant

**Comment by sylee957** [comment]
I can merge this
But please add some tests and follow the PEP-8 guidelines and community code quality guidelines.

**Comment by abhinav28071999** [comment]
So `_verify_invertible(M, iszerofunc=iszerofunc)` sould be removed from every matrix inversion method ?

**Comment by sylee957** [comment]
If you think so, you can remove them
But I can't guarantee that strange test failings can happen after removing them.

File: sympy/matrices/inverse.py [hunk 1]
```diff
@@ -316,6 +316,44 @@ def _inv_QR(M, iszerofunc=_iszero):
...
+    inverse_GE
+    inverse_CH
+    inverse_LDL
+    """
+    from sympy import BlockMatrix
+    i = M.shape[0]
+    if i <= 20 :
+        return M.inv(method="LU", iszerofunc=_iszero)
+    A = M[:i // 2, :i //2]
+    B= M[:i // 2, i // 2:]
```
Comment (sylee957) [inline sympy/matrices/inverse.py]:
```suggestion
    B = M[:i // 2, i // 2:]
```

File: sympy/matrices/inverse.py [hunk 2]
```diff
@@ -316,6 +316,44 @@ def _inv_QR(M, iszerofunc=_iszero):
...
+    C = M[i // 2:, :i // 2]
+    D = M[i // 2:, i // 2:]
+    try:
+        D_inv = _inv_BLOCK(D)
+    except NonInvertibleMatrixError:
+        return M.inv(method="LU", iszerofunc=_iszero)
+    B_D_i = B* D_inv
```
Comment (sylee957) [inline sympy/matrices/inverse.py]:
```suggestion
    B_D_i = B * D_inv
```

**Comment by abhinav28071999** [comment]
@sylee957  we can improve of inverse by :
1. I think it is possible to check if the matrix is invertible or not inside the decomposition algo
2. We can use strassen algo to compute matrix product. Here when computing inv of 200*200 matrix,
6 100*100 matrix multiplication, 12 50*50 matrix multiplication are involved, it can be made efficient with strassen algo

**Comment by sylee957** [comment]
If you want to try out some ideas further, you can try it.
But I have given up lots of attempts, because of the sympy overhead diminishing the advantage, or even making such block formulations worse than naive multiplications.
For example, creating a new matrix can be slow right now.

**Comment by oscarbenjamin** [comment]
See also #16207 which shows a way to permute rows and columns to make a matrix block diagonal.

**Comment by smichr** [comment]
This seems to work well for sparse matrices:
```python
>>> from time import time
>>> a=randMatrix(20,percent=40)
>>> t=time();b=_inv_BLOCK(a);time()-t
16.804502725601196
>>> assert a*b == eye(20)
>>> t=time(); ok=a.inv();time()-t
159.12335085868835
```

**Comment by abhinav28071999** [comment]
> This seems to work well for sparse matrices:


The performance can be improved further if inv using LU can return without checking for determinant.

**Comment by abhinav28071999** [comment]
`    try:`
     `   iv =  M.LUsolve(M.eye(M.rows), iszerofunc=_iszero)`
    `    if iv * M == M.eye(M.rows):`
   `         return iv`
  `      else:`
 `           raise NonInvertibleMatrixError("Matrix det == 0; not invertible.")`
`    except:`
        raise NonInvertibleMatrixError("Matrix det == 0; not invertible.")`
`
 Will this work in `_inv_LU`

**Comment by smichr** [comment]
> if iv * M == M.eye(M.rows):

You might want to test this as `all(p.is_diagonal() and all(i==1 for i in p.diagonal(0)) for p in [iv*M])` so you don't have to create another matrix.

**Comment by sylee957** [comment]
There is already is_Identity for use

**Comment by smichr** [comment]
> There is already is_Identity for use

I tried `.is_identity` -- I should have looked for `[i for i in dir(Matrix) if 'dent' in i]`

**Comment by abhinav28071999** [comment]
@smichr @sylee957 Something like this might work.

**Comment by abhinav28071999** [comment]
A better way might be :
`    try:`
`        A = M.LUsolve(M.eye(M.rows), iszerofunc=_iszero)`
 `   except (NotImplementedError, ValueError):`      
    `    raise NonInvertibleMatrixError("Matrix det == 0; not invertible.")`
 `       return A`

**Comment by oscarbenjamin** [comment]
> `except (NotImplementedError, ValueError):`

Why do you need to catch those exceptions? It would be better not to catch any exceptions.

If there is a good reason for not knowing in advance whether those exceptions would be raised then LUsolve should be changed to raise something more specific like `NonInvertibleMatrixError`. Then you wouldn't need to catch the exception here.

**Comment by abhinav28071999** [comment]
@oscarbenjamin Yes you are right, the exception in `_LUsolve` should be changed.

**Comment by abhinav28071999** [comment]
@oscarbenjamin If i  change the Exception name in _LU_solve it give error in several test cases.
When LU_solve calls _LUsolve then Not_Implemented is expected and when inv call then Not_Invertable is expected.

**Comment by oscarbenjamin** [comment]
> @oscarbenjamin If i change the Exception name in _LU_solve it give error in several test cases.
> When LU_solve calls _LUsolve then Not_Implemented is expected and when inv call then Not_Invertable is expected.

There might need to be changes in other places.

**Comment by abhinav28071999** [comment]
Is it okay to implement faster inversion with determinant check in decomposition method  in a separate pr.  This pr should be implementing only block inversion method.

File: sympy/matrices/inverse.py [hunk 3]
```diff
@@ -316,6 +314,44 @@ def _inv_QR(M, iszerofunc=_iszero):
 
     return M.QRsolve(M.eye(M.rows))
 
+def _inv_BLOCK(M, iszerofunc=_iszero):
```
Comment (oscarbenjamin) [inline sympy/matrices/inverse.py]: This should be called `_inv_block`

Comment (abhinav28071999) [reply sympy/matrices/inverse.py]: Done.

**Comment by abhinav28071999** [comment]
I will remove the determinant check from inversion methods in a separate pr.

**Comment by sylee957** [comment]
Since this is adding a new keyword, I may only let to hear what should this be named for.
'BLOCK', 'block'?

**Comment by smichr** [comment]
> what should this be named

in keeping with CAPS abbreviations, how about BLK? though that looks a little like BLACK or BULK but in context it should be clear. Or we could use the suggestive '[x]' in spirit of `oo` that is used for "infinity".

**Comment by sylee957** [comment]
Whatever, I'd merge this because I can't think of any naming that is 'ultimate'.

I think that block inversion should be operating on a completely different level than the methods like `LU`, `QR` because it's always possible to swap such methods internally, but this sort of consideration would be done after many other block inversion formulas are introduced.

**Comment by oscarbenjamin** [comment]
I agree that blocking should be handled at a different level. I also think #16207 should be implemented for matrix inversion as part of that. The technique there naturally handles the diagonal case and massively improves performance for wide classes of sparse matrices.

Does the block inversion approach here change the order of the elementary arithmetic? We need to bear in mind that elementary arithmetic laws are not always interchangeable in sympy so that e.g.
```
In [11]: a, b, c = symbols('a, b, c')                                                                                                          

In [12]: (a+b)/c == a/c + b/c                                                                                                                  
Out[12]: False
```

**Pull request merged**